streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+it needs. You can see that here in the `shop3.js` script, which includes the BTP-enabled ledger; run `diff shop2.js shop3.js`
+to see how similar they really are:
+
+```js
+const IlpPacket = require('ilp-packet')
+const http = require('http')
+const crypto = require('crypto')
+const Plugin = require('ilp-plugin-payment-channel-framework')
+function base64 (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+function hash (secret) { return crypto.createHash('sha256').update(secret).digest() }
+function hmac (secret, input) { return crypto.createHmac('sha256', secret).update(input).digest() }
+
+let users = {}
+const store = {}
+const plugin = new Plugin({
+  listener: {
+    port: 9000
+  },
+  incomingSecret: '',
+  maxBalance: '1000000000',
+  prefix: 'g.letter-shop.mytrustline.',
+  info: {
+    currencyScale: 9,
+    currencyCode: 'XRP',
+    prefix: 'g.letter-shop.mytrustline.',
+    connectors: []
+  },
+  _store: {
+    get: (k) => store[k],
+    put: (k, v) => { store[k] = v },
+    del: (k) => delete store[k]
+  }
+})
+
+plugin.connect().then(function () {
+  plugin.on('incoming_prepare', function (transfer) {
+    const ilpPacketContents = IlpPacket.deserializeIlpPayment(Buffer.from(transfer.ilp, 'base64'))
+    const parts = ilpPacketContents.account.split('.')
+    // 0: test, 1: crypto, 2: xrp, 3: rrhnXcox5bEmZfJCHzPxajUtwdt772zrCW, 4: userId, 5: paymentId
+    if (parts.length < 6 || typeof users[parts[4]] === 'undefined' || ilpPacketContents.amount !== transfer.amount) {
+      plugin.rejectIncomingTransfer(transfer.id, {}).catch(function () {})
+    } else {
+      const { secret, res } = users[parts[4]]
+      const fulfillment = hmac(secret, ilpPacketContents.account)
+      const condition = hash(fulfillment)
+      if (transfer.executionCondition === base64(condition)) {
+        plugin.fulfillCondition(transfer.id, base64(fulfillment)).then(function () {
+          const letter = ('ABCDEFGHIJKLMNOPQRSTUVWXYZ').split('')[(Math.floor(Math.random() * 26))]
+          res.write(letter)
+        }, function (err) {
+          console.error(err.message)
+        })
+      } else {
+        console.log('no match!', { secret, fulfillment, condition, transfer })
+      }
+    }
+  })
+
+  http.createServer(function (req, res) {
+    const secret = crypto.randomBytes(32)
+    const user = base64(crypto.randomBytes(8))
+    users[user] = { secret, res }
+    console.log('user! writing head', user)
+    res.writeHead(200, {
+      'Pay': [ 1, plugin.getAccount() + '.' + user, base64(secret) ].join(' ')
+    })
+    // Flush the headers in a first TCP packet:
+    res.socket.write(res._header)
+    res._headerSent = true
+  }).listen(8000)
+})
+```
+
+To run it, use:
+
+```sh
+npm install interledgerjs/ilp-plugin-payment-channel-framework#mj-fix
+node ./shop3.js
+```
+
+And here is the content consumption tool, which first transfers 1000 XRP to its account at the shop's ledger, and then
```

Comment:
I think someone might reasonably be confused about why it's better to transfer all the money to the shop before they send you anything than to do it in exchange for the letters. I think BTP makes more sense when you're funding an account with a connector that you'll use for a lot of payments to different receivers.

Reply:
Ah, actually that text is incorrect, that's what I wanted to do first, but I ended up just sending the drops from a `setInterval`. But that's still putting a lot of trust in the shop; I could change the code so that it counts how many letters were received, and limits the in-flight amount (e.g. if the shop is behind by more than 10 letters, pause the money sending for a while)